BANNO.M2MTRANSFERS.V3.POW Xacct fix

tkainz opened the pull request:
**   Ver. 3.2.1  11/06/2023 T. Kainz -
 * Corrected issue where cross-account loans were being allowed
 when only shares were requested.
 * Updated last name verification to properly process last names stored with one or more spaces at the beginning.
 * Adjusted CFG letter file param logic to accept both "LOAN" and "LOANS" for the eligible types of transfer from shares

jordan-keenan commented on m2m-transfers/V3/REPWRITERSPECS/BNO.INTERACTIVE.PRO:
```diff
@@ -250,12 +258,16 @@ PROCEDURE NTADDCROSSACCOUNTS
+       BNODEBUGMSG=FORMAT("** NTSLTYPE:99:",NTSLTYPE)
+       CALL BNODEBUGMSGADD
+       IF (NTPREFACCESSIDTYPE=2 OR NTPREFACCESSIDTYPE=0) AND
+          (NTCROSSACCTSLTYPE=NTSLTYPESHARE OR
```
Hi Tom, Is this fix actually a Bug in BNO.INTERACTIVE.PRO that is an improvement for any other PO's that may use that functionality in that include?

tkainz replied:
@jordan-keenan , yes, the issue is resolved by a change in the ".PRO"

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BNO.INTERACTIVE.PRO:
```diff
@@ -250,12 +258,16 @@ PROCEDURE NTADDCROSSACCOUNTS
+       CALL BNODEBUGMSGADD
+       IF (NTPREFACCESSIDTYPE=2 OR NTPREFACCESSIDTYPE=0) AND
+          (NTCROSSACCTSLTYPE=NTSLTYPESHARE OR
+           NTCROSSACCTSLTYPE=NTSLTYPESHARELOAN) THEN
```
Do we also need to have a check for NTSLTYPESHARELOAN up above starting at line #226?

tkainz replied:
@rrobison44 is this what you're referring to - the reference to unrestricted ID Types (line 231)?

![image](https://github.com/Banno/banno-powerons/assets/60237061/d9ab3f7c-b8bf-49dc-a61c-effa2468de4d)

rrobison44 reviewed (commented):
Added some comments.

tkainz commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -2317,20 +2325,20 @@ PROCEDURE READCONFIGFILESETTINGS
                PARAMVALUE<>"" THEN
         DO
-         IF UPPERCASE(PARAMVALUE)="LOAN" THEN
+         IF UPPERCASE(PARAMVALUE)="LOAN" OR
```
@rrobison44 Since the default (per the CFG Letter file) is "SHARES", I feel that falling back on that value as the default is the "safest" bet to make.  If the CFG Letter file setting is not either (specifically) "LOAN", "LOANS" or "BOTH" then the program falls back on to "SHARES". The alternative would be to add specific tests for "SHARE" and "SHARES" but then we'd have to add in additional error handling in the event it wasn't ONLY set up as "SHARE", "SHARES", "LOAN", "LOANS" or "BOTH".

rrobison44 replied:
Ok.  I guess I was thinking you could still just fall back to shares if the value is not in the list: "SHARE", "SHARES", "LOAN", "LOANS" or "BOTH".  I think that is what you are doing now, anyways.  A value of "SOMETHING" would fall back to shares now.  My only point is that it is more clear if you actually have the "SHARE"/"SHARES" checks in the config read.  But it is good as is...in any case.

tkainz replied:
@rrobison44 - Would you be thinking of something like this:

`
         IF UPPERCASE(PARAMVALUE)="LOAN" OR
            UPPERCASE(PARAMVALUE)="LOANS" THEN
          PARAMNTSLTYPE=NTSLTYPELOAN
         ELSE IF UPPERCASE(PARAMVALUE)="SHARE" OR
              UPPERCASE(PARAMVALUE)="SHARES" THEN
           PARAMNTSLTYPE=NTSLTYPESHARE
         ELSE IF UPPERCASE(PARAMVALUE)="BOTH" THEN
          PARAMNTSLTYPE=NTSLTYPESHARELOAN
         ELSE
          PARAMNTSLTYPE=NTSLTYPESHARE
`

tkainz replied:
@rrobison44 Requested change committed and pushed.

tkainz commented on m2m-transfers/V3/REPWRITERSPECS/BNO.INTERACTIVE.PRO:
```diff
@@ -3,6 +3,11 @@
 |
+|  Modified By    : TKainz
+|  Modified Date  : Nov. 06, 2023
+|  Modification   : Corrected issue where cross-account loans were being
```
@rrobison44 - Not sure, actually - didn't specifically test for that - before this fix. With the fix installed, I did test all three values ("SHARES", "LOANS", "BOTH") with the results being returned as expected.

rrobison44 replied:
ok.

jucarson reviewed (approved):
Looks good Tom!

tkainz commented:
- Added the inclusion of the new Share/Loan list build procedures to the program, removing the old NT code. 
- Added the removal of leading spaces from the last name (if any) as these were causing the validation to fail.
- Added check and enforcement of the tran comment length for immediate transfers to keep the transaction from failing if they were too long.

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -1981,21 +1967,36 @@ PROCEDURE CALCLIMITS
+** funds from
+** 3.2.1 - Updated procedures being called
+*]
+ CALL BNOSETUSERHBENABLE
```
Isn't BNOSETUSERHBENABLE already called in BNOBUILDSHARELOANLISTINIT?

tkainz replied:
Yep - removed

rrobison44 reviewed (changes requested):
Just added a few comments and requested a couple changes.  Also, just to confirm, the BNOLB files you are including with this PR are just the ones from your earlier PR that has already been reviewed??  So no need to review those, correct?

tkainz commented:
@rrobison44 @jucarson @jordan-keenan .... just to clarify... the "Includes" in this PR are the ones you just recently reviewed as stand-alone includes so there really shouldn't be any need to review them as part of this PR.... just to the extent as how they're being called and initialized in the M2M program itself.

tkainz commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -1471,92 +1451,99 @@ PROCEDURE PRINTCURRENTSTATE
    NEWLINE
-   PRINT INDENT(BASEINDENT+4)+Q+"accountOwnerName"+Q+": "+Q+NTLOANACCOUNTNAME(SLLOOP)+Q+","
+   PRINT INDENT(BASEINDENT+4)+Q+"accountOwnerName"+Q+": "+Q+
+         BNOSHAREXACCTPRIMARYNAME(SLLOOP)+Q+","
```
Corrected

tkainz commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -2403,10 +2406,12 @@ PROCEDURE READCONFIGFILESETTINGS
           PARAMIDLENTEST=TESTIDLENGTH4
         END
 [* Default Transaction Comment
+** 3.2.1 - Limit default comment to 19 characters to avoid possibility
```
You are correct - this just limits what the program is grabbing from the parameter settings so, when combined with the member's input, the limit will not be exceeded.  Comment removed.

tkainz commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -2730,21 +2735,21 @@ PROCEDURE ERRORHANDLER
  IF ERRORDETAILOFFSET=4 THEN
   DO
-   ERRORDETAIL(2)=ERRORDETAIL(2)+"Cannot calc member limits."
+   ERRORDETAIL(4)=ERRORDETAIL(4)+"Cannot calc member limits."
```
Done

tkainz commented:
@rrobison44 your recent peer review notes have been implemented and pushed.

rrobison44 reviewed (approved):
Looks great!  Thanks, Tom.

jucarson reviewed (approved):
Changes look great. Thanks Tom!

jordan-keenan reviewed (approved):
The changes look great, Tom!
Thank you for cleaning up and making a new set of Include Files that we can use
for Banno PowerOns that have the variables and procedures  staring with BNO now :)

jordan-keenan reviewed (commented):
Hi Tom,
 The changes from 01/08/24 look good.
This goes without saying, just be sure to add the SLLISTBUILD .DEF and .PRO updates for BNOINCLUDECLOSED and BNOINCLUDECHARGEDOFF when you include new changes to handle quotes surrounding the Share descriptions in Symitar to work with the JSon.

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -1171,6 +1177,9 @@ PROCEDURE GETRECIPSLDATA
        NEWTRANSLDESCRIPTION=SHARE:NICKNAME
        IF NEWTRANSLDESCRIPTION="" THEN
         NEWTRANSLDESCRIPTION=SHARE:DESCRIPTION
+       CHARCHECK=NEWTRANSLDESCRIPTION
```
Instead of doing the checking/scrubbing here, can we just do the check where we are outputting the json?

tkainz replied:
@rrobison44 : removed

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -1181,10 +1190,13 @@ PROCEDURE GETRECIPSLDATA
        NEWTRANSLDESCRIPTION=LOAN:NICKNAME
        IF NEWTRANSLDESCRIPTION="" THEN
         NEWTRANSLDESCRIPTION=LOAN:DESCRIPTION
+       CHARCHECK=NEWTRANSLDESCRIPTION
```
Instead of doing the checking/scrubbing here, can we just do the check where we are outputting the json?

tkainz replied:
@rrobison44 : Removed

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -2070,10 +2088,13 @@ PROCEDURE BUILDSLTRANSFERLIST
          TRANACCOUNTNAME(TRANCOUNT)=SHARE:NICKNAME
          IF SHARE:NICKNAME="" THEN
           TRANACCOUNTNAME(TRANCOUNT)=SHARE:DESCRIPTION
+         CHARCHECK=TRANACCOUNTNAME(TRANCOUNT)
```
Instead of doing the checking/scrubbing here, can we just do the check where we are outputting the json?

tkainz replied:
@rrobison44 : Removed

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -3008,6 +3029,9 @@ PROCEDURE PULLRGDATA
            NEWTRANSLDESCRIPTION=SHARE:NICKNAME
            IF NEWTRANSLDESCRIPTION="" THEN
             NEWTRANSLDESCRIPTION=SHARE:DESCRIPTION
+           CHARCHECK=NEWTRANSLDESCRIPTION
```
Instead of doing the checking/scrubbing here, can we just do the check where we are outputting the json?

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -3034,6 +3058,9 @@ PROCEDURE PULLRGDATA
            NEWTRANSLDESCRIPTION=LOAN:NICKNAME
            IF NEWTRANSLDESCRIPTION="" THEN
             NEWTRANSLDESCRIPTION=LOAN:DESCRIPTION
+           CHARCHECK=NEWTRANSLDESCRIPTION
```
Instead of doing the checking/scrubbing here, can we just do the check where we are outputting the json?

tkainz replied:
@rrobison44 : Removed

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -3136,6 +3163,9 @@ PROCEDURE PULLRGDATA
          NEWTRANSLDESCRIPTION=SHARE:NICKNAME
          IF NEWTRANSLDESCRIPTION="" THEN
           NEWTRANSLDESCRIPTION=SHARE:DESCRIPTION
+         CHARCHECK=NEWTRANSLDESCRIPTION
```
Instead of doing the checking/scrubbing here, can we just do the check where we are outputting the json?

tkainz replied:
@rrobison44 : Removed

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -3282,6 +3312,28 @@ PROCEDURE NORMALIZESLID
   END
 END [PROCEDURE]
 
```
I'm thinking the CHARCHECKTMP variable is not necessary.  And maybe just do the SEGMENT one time to get the current character that is processing:
```
PROCEDURE CHECKFORINVALIDCHARS
[* Check the passed character value for any double-quotes (ascii 34)
** and if any found replace with a single-quote character (ascii 39).
**  PASS    CHARCHECK  Character value being tested
**  RETURN  CHARCHECKRETURN  Updated CHARCHECK character value
*]
 CHARCHECKRETURN=""

 NLTSCHR=CHARCHECK
 CALL NLTS
 CHARCHECK=NLTSCHR

 FOR CHARCHECKNUM=1 TO LENGTH(CHARCHECK)
  DO
   TMPCHR=SEGMENT(CHARCHECK,CHARCHECKNUM,CHARCHECKNUM)
   IF TMPCHR=CTRLCHR(34) THEN
    CHARCHECKRETURN=CHARCHECKRETURN+CTRLCHR(39)
   ELSE
    CHARCHECKRETURN=CHARCHECKRETURN+TMPCHR
  END
END [PROCEDURE]
```

tkainz replied:
@rrobison44 Updated as (almost) requested

rrobison44 reviewed (changes requested):
See comments.  Generally, I think we should just be calling CHECKFORINVALIDCHARS when we our creating json output.  That is done in some cases but in other cases it is done when collecting data.  I think the code get simplified if you just do the check in the json output sections of code.

rrobison44 commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -1591,16 +1591,16 @@ PROCEDURE PRINTCURRENTSTATE
-
+   CHARCHECK=TRANACCOUNTNAME(TMPLOOP)
+   CALL CHECKFORINVALIDCHARS
+   TRANACCOUNTNAME(TMPLOOP)=CHARCHECKRETURN
```
Maybe remove this line where you are assigning CHARCHECKRETURN back to the TRANACCOUNTNAME array?  Instead just reference CHARCHECKRETURN on the next line where you are printing the `accountName` property.

tkainz replied:
@rrobison44 : I was actually thinking about that but I figured that updating the same variable with the corrected/modified value was the safest bet as otherwise, I'd have to go through the code after printing the value to verify that leaving the original (uncorrected/unmodified) variable value intact wouldn't be causing further issues downstream.

rrobison44 reviewed (approved):
Looks good, Tom.  Just one more suggestion but I'm approving as is.  Thanks.

jordan-keenan commented on m2m-transfers/V3/REPWRITERSPECS/BANNO.M2MTRANSFERS.V3.POW:
```diff
@@ -1998,6 +1998,8 @@ PROCEDURE BUILDSHARELOANLIST
  BNOSLCODE(BNOSHARECODECLUB)=TRUE
```
The code looks good. Found some clean up items from before.
On Line 386 in the Set up division please correct the following spelling:
"[*the followowing freq refreences are'

tkainz replied:
@jordan-keenan Corrected.

jordan-keenan replied:
@tkainz Hi Tom, For M2M thanks for improving it so much, and for fixing the typo from a long time ago.
I noticed that sentence that had the typo corrected still has a typo spelling "following" as "followowing" Can you fix that, then I'll approve the review?
"the followowing freq references"
Thanks,
Jordan

jordan-keenan replied:
Thank you @tkainz ! I see that you've changed that, and I've approved the code.

jordan-keenan reviewed (approved):
The M2M changes look great, including changing the Share/Loan Build List to use Banno parameter names, etc.! Thank you for doing all that Tom :)
